sys/compat/linuxkpi/common/src/linux_compat.c

```diff
@@ -824,44 +826,44 @@ linux_remap_address(void **uaddr, size_t len)
 		uaddr_val += (uintptr_t)pts->bsd_ioctl_data;
 
 		/* update address location */
-		*uaddr = (void *)uaddr_val;
+		*uaddr = (void * __capability)(uintcap_t)uaddr_val;
 		return (1);
 	}
 	return (0);
 }
 
 int
-linux_copyin(const void *uaddr, void *kaddr, size_t len)
+linux_copyin(const void * __capability uaddr, void *kaddr, size_t len)
 {
-	if (linux_remap_address(__DECONST(void **, &uaddr), len)) {
+	if (linux_remap_address(__DECONST(void * __capability *, &uaddr), len)) {
 		if (uaddr == NULL)
 			return (-EFAULT);
-		memcpy(kaddr, uaddr, len);
+		memcpy(kaddr, (void *)(__cheri_addr uintptr_t)uaddr, len);
 		return (0);
 	}
 	return (-copyin(uaddr, kaddr, len));
 }
 
 int
-linux_copyout(const void *kaddr, void *uaddr, size_t len)
+linux_copyout(const void *kaddr, void * __capability uaddr, size_t len)
 {
 	if (linux_remap_address(&uaddr, len)) {
 		if (uaddr == NULL)
 			return (-EFAULT);
-		memcpy(uaddr, kaddr, len);
+		memcpy((void *)(__cheri_addr uintptr_t)uaddr, kaddr, len);
 		return (0);
 	}
 	return (-copyout(kaddr, uaddr, len));
 }
 
 size_t
-linux_clear_user(void *_uaddr, size_t _len)
+linux_clear_user(void * __capability _uaddr, size_t _len)
 {
-	uint8_t *uaddr = _uaddr;
+	uint8_t * __capability uaddr = _uaddr;
 	size_t len = _len;
 
 	/* make sure uaddr is aligned before going into the fast loop */
-	while (((uintptr_t)uaddr & 7) != 0 && len > 7) {
+	while (((__cheri_addr uintptr_t)uaddr & 7) != 0 && len > 7) {
```

Comment:
Probably a good place for `__builtin_is_aligned()`. I have an `is_aligned()` macro in the purecap kernel branch that I should probably have merged sooner. We could use the builtin here and then move it to the macro when we merge it.

Reply:
Do we have wrapper macros for __builtin_is_aligned? I think we changed roundup2 and rounddown2 to use the builtins but not sure about this one.

Reply:
We didn't change roundup/rounddown because they now DTRT as-is.  We had explicitly used the builtin in some places that might have gone back to using round* to minimize diffs with upstream.

Reply:
I did end up using `__builtin_is_aligned` here.